### ISSUE TITLE
Improve scalability through supporting multiple exec servers

### ISSUE THREAD

**Issue by hluu** [azkaban/azkaban#439]
## Current Challenge
Azkaban consists of two servers - web server and executor server.  The current design is the web server assigns flow executions to a single executor server.  Web server mainly provides the UI and the scheduling features, and executor server is responsible for running the jobs in the flow.  Azkaban is running into a scalability issue due to the increased usage as well as user jobs are taking longer and longer to finish because of our large ever increasing data sets. Given the increased number of jobs the executor server has to execute concurrently, it is running into system memory capacity issue and therefore the server hosting Azkaban web server and executor is busy with performing swapping and the web server response is dramatically slow.

## Proposed Design
At the high level, Azkaban server scalability can be improvied by deploying the web server on its own box and deploying multiple executor servers on separate boxes .  With this enhancement, the job executions load can spread across multiple machines and therefore Azkaban server can support more load and can tolerate server failure.
Since Azkaban web server controls the scheduler and there is only one web server, it can still guarantee that a single job is executed only once, even though there are multiple executor servers.
The implementation of this design should make sure it is still easy and possible to run Azkaban in single-server mode, which is very useful for development and debugging purposes.

## Enhancements to Web Server
Accept a list of one or more executor servers - i.e <server name>:<port>.  This information should be backed by a database table so web server doesn't have to be bounced when this list is modified.
When executing a flow, web server needs to decide which executor to send a job execution request to. The logic for the load balancing algorithm should be pluggable.  We will provide a default implementation out of the box.
The JMX page needs to be enhanced to show which flows are being executed on which executor server as well as their current states.
The web server may need to adjust its internal thread pool based # of running executors.  This way, as we scale down from 5 to 1 executor servers, the web server doesn't overwhelm the one executor with flows to execute.
Need a new property to specify the implementation of the pluggable load balancing logic.
When it is time to figure out which executor to execute a flow, web server will delegate that responsibility to the load balancer.

## Enhancements to Executor Server
Executor server should support various states - INITIALIZING, ACCEPTING, NOT_ACCEPTING, SHUTTING_DOWN.  May need to come up with better name for these states.
INITIALIZING - while server is coming up.  It doesn't accept requests
ACCEPTING - up and running, accepting requests
NOT_ACCEPTING - up and running, but not accepting requests.  Good for draining out the currently running flows as we are doing rolling upgrade.
SHUTTING_DOWN - coming down, not accepting requests. (not sure if we need this)
Need a new REST endpoint to change the state to NOT_ACCEPTING or ACCEPTING
While the server is INITIALIZING or SHUTTING_DOWN state, its state can't be changed
Need a new REST endpoint to return the current load information (how many flows are currently being executed)

## Flow Execution Load Balancer
The goal of this load balancer is to distribute the flow executions evenly across 1 or more executor servers
We can start with the basic logic by picking the executor server with the least amount of load
Only executor servers that are in the "ACCEPTING" state should be considered for a candidate for executing a flow
If an error while connecting to the selected executor server, the next best candidate is selected.
There should be a background thread to periodically ping the list of executors for their current status and update that information in memory.

**Comment by RexGibson** [comment]
Much work will contain state and local computation on the Azkaban server. I would suggest that a solution for moving that state is considered if you are going to distribute a flow across multiple executors.  Alternatively I would accept a simpler solution that allowed a flow to be pinned to a specific server and all work for that flow is located there. I would love it if you could have kinds of executor servers akin to jenkins slaves where you could pin work to a note that it optimized for the work needed in the flow. I think a holy grail would also have a plugin that could spin up an new executor and register it with the "cluster".

**Comment by RexGibson** [comment]
However I was under the impression that Azkaban wasn't going to have the manpower investment made by LinkedIn to solve the executor problem because they are going to Ooozie.  We are currently considering other Workflow managers because of this understanding that multiple executors would not be invested in.  I would love some sort of clarity as to if this project is being carried forward by LinkedIn in a committed way or if this has become a lower priority.

**Comment by hluu** [comment]
Another interesting thought would be to spin up a Yarn application to manage the life cycle of a flow, which involves executing all the jobs in that flow.

**Comment by RexGibson** [comment]
That doesn't have as much value to us as we are not using azkaban as a direct hadoop connector we are firing off fruit fly clusters in EMR via Azkaban.

**Comment by ptalathi** [comment]
Thanks a lot for the post.

@Sokrati (http://www.sokrati.com/), we are using Azkaban as our internal workflow manager and have been planning to implement multiple executor support. Here's a brief design description. Any suggestions/comments are most welcome. You may contact us at [farnoosh.azadi@, gaurav.shaha@, prachita.talathi@]sokrati.com.


Table: executors
Primary key: host + port
Name		  Datatype	NULLable
host			  varchar   	no		
port			  int		        no
load			  int		        yes
heartbeat		  bigint	        yes
status_type_id	  int		        no



Table: server_status_types
Primary key: id
Name		Datatype	       NULLable
id			int		       no		
name		varchar	       no
is_deleted	tinyint	       no		

id			name			is_deleted
1 			INITIALISING		0
2			RUNNING		0
3			STANDBY		0
4			DOWN			0

azkaban.properties for executor server:
As the executor server breaks after the no. of simultaneous executions pass 120-130, set the executor.flow.threads value in config to a safe limit of 100 for all executors. Set the executor.port to chosen port number.

azkaban.properties for web server:
As the executor host & port information will be read directly from database, no need to mention those in the conf file. Add a new property for load.balancer as well as loading.factor.

REST endpoints:

Alter the status of an executor: 
Permissible: RUNNING->STANDBY, STANDBY->RUNNING

Method: GET
Request URL: /manager?ajax=alterStatus
Parameter Location: Request Query String

Parameter              Description
session.id               The user session id.
ajax=alterStatus     The fixed parameter indicating the current ajax action
executorHost          IP address of the executor server
executorPort           Port number of the executor server
status                      String indicating new status 

AzkabanExecutorServer:
AzkabanExecutorServer():
Add an entry to executors table with host, port information in not present, else mark the status as INITIALISING.
stopServer():
Mark the status in executors table as DOWN.

ExecutorManager:
submitExecutableFlow():
Upon receipt of an executeFlow request, ask the load balancer to pick and send the executeFlow request to it.
if failed to submit request, notify load balancer and ask to pick another executor.
finalizeFlows():
For a finished(failed/successful) job, decrement the load count of its executor.

LoadBalancer:
numRunningExecutors 
loadingFactor
maxThreads = loadingFactor * numRunningExecutors * executorFlowThreads
selectExecutor():
Pick the least loaded executor in RUNNING state. (Load = number of running flows)
Send the executorReference.
submissionSuccessful(executionReference):
Increment the load count for given executor.
submissionFailed(attempts):
select the (n+1)th least loaded executor where n = attempts.
getExecutorStatus(executionReference):
Return executor status
setExecutorStatus(executionReference, status):
Set executor status


executorWatchdogThread:
obtain the ListofExecutors from executors table
run()
for entry in ListofExecutors
if status == INITIALISING, Create an updaterThread and change the status to RUNNING.
else if status == RUNNING, ping executor for status
if no response, update status as DOWN.
sleep;

ExecutingManagerUpdaterThread: 
Every executor will have its own updaterThread.
run()
if running == true
list = list of all running flows on executor
if getExecutorStatus() == RUNNING
Fetch flow list for the executor
for flow: list
Fetch update
if failed/successful, add to finishedFlows
if executor unresponsive even after retries, notify ExecutorWatchdogThread.
else if getExecutorStatus() == DOWN
for flow:list
DELETE entry from activeExecutingFlows table, UPDATE status as UNREACHABLE in execution_flows table.
running = false

Status:
Add UNREACHABLE in job status types.

JMX:
Add multiple executor support in JMX pages.

**Comment by winghc** [comment]
This is the solution what we are looking for. Like Hadoop Map/Reduce, there can be many mappers(executors),  when designing job workflow, one can assign which executor to run in order to balance the executors.   As RexGibson mentioned, is linkedin going to move manpower to oozie?

**Comment by BrianGallew** [comment]
Shouldn't this be closed now?

**Comment by logiclord** [comment]
yes
